datanode/datanode.go

```diff
@@ -303,39 +327,52 @@ func (d *dataNode) startAnalyzingShardAvailability() {
 
 		dynamicTopo, ok := d.topo.(topology.DynamicTopology)
 		if !ok {
-			d.logger.Error("cannot mark shard available", zap.Error(errors.New("topology is not dynamic")))
+			d.logger.Error("cannot mark shard available, topology is not dynamic")
 			continue
 		}
 
-		tables := d.Tables()
 		availableShards := make([]uint32, 0, len(initializing))
+		// snapshot fact tables
+		// map from table name to whether the table is a fact table
+		factTables := make([]string, 0)
+		d.memStore.RLock()
+		for table, schema := range d.memStore.GetSchemas() {
+			if schema.Schema.IsFactTable {
+				factTables = append(factTables, table)
+			}
+		}
+		d.memStore.RUnlock()
+
 		// check all whether all tables within initialing shards are bootstrapped
 		for shardID := range initializing {
 			numTablesBootstrapped := 0
-			for _, table := range tables {
+			for _, table := range factTables {
```

Comment:
who marks dimension table availability?

Reply:
under current m3 framework we don't have availability tracking for dimension table (non sharded) out of box. we will need to add dimension table availability tracking ourselves in etcd

Reply:
I see. maybe add a todo somewhere